flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -930,6 +943,10 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
 
         InputChannel inputChannel = inputChannelsWithData.poll();
         enqueuedInputChannelsWithData.clear(inputChannel.getChannelIndex());
+        checkState(
+                !channelsWithEndOfPartitionEvents.get(inputChannel.getChannelIndex()),
+                "Channel already received EndOfPartition %s",
+                inputChannel.getChannelInfo());
```

Comment:
The checkstate feels a bit overkill now, but as long as there is no impact on performance, I guess it's better to be safe than sorry.

Reply:
I removed the check as there was degradation in some benchmarks.